File: backend/fastapi/app/core/cache.py
"""
Redis缓存操作核心模块
"""

import json
from typing import Dict, Optional
import redis.asyncio as redis

from app.config import settings


class CacheManager:
    """Redis缓存管理器"""
    
    def __init__(self):
        self.redis_client: Optional[redis.Redis] = None
    
    async def init_redis(self):
        """初始化Redis连接"""
        self.redis_client = redis.from_url(settings.REDIS_URL, decode_responses=True)
        
    async def close_redis(self):
        """关闭Redis连接"""
        if self.redis_client:
            await self.redis_client.close()
    
    async def get_cached_data(self, key: str) -> Optional[Dict]:
        """从Redis获取缓存数据"""
        try:
            if not self.redis_client:
                return None
            data = await self.redis_client.get(key)
            return json.loads(data) if data else None
        except Exception as e:
            print(f"Redis读取失败: {e}")
            return None

    async def set_cached_data(self, key: str, data: Dict, ttl: int = None):
        """设置Redis缓存数据"""
        try:
            if not self.redis_client:
                return
            
            ttl = ttl or settings.CACHE_TTL
            await self.redis_client.setex(key, ttl, json.dumps(data))
        except Exception as e:
            print(f"Redis写入失败: {e}")
    
    async def delete_cache(self, key: str):
        """删除缓存"""
        try:
            if not self.redis_client:
                return
            await self.redis_client.delete(key)
        except Exception as e:
            print(f"Redis删除失败: {e}")
    
    async def delete_cache_pattern(self, pattern: str):
        """批量删除匹配模式的缓存"""
        try:
            if not self.redis_client:
                return
            keys = await self.redis_client.keys(pattern)
            if keys:
                await self.redis_client.delete(*keys)
        except Exception as e:
            print(f"Redis批量删除失败: {e}")
    
    async def ping(self) -> bool:
        """检查Redis连接状态"""
        try:
            if not self.redis_client:
                return False
            await self.redis_client.ping()
            return True
        except Exception:
            return False
    
    async def get_info(self, section: str = "memory") -> Dict:
        """获取Redis信息"""
        try:
            if not self.redis_client:
                return {}
            return await self.redis_client.info(section)
        except Exception as e:
            print(f"获取Redis信息失败: {e}")
            return {}


# 全局缓存管理器实例
cache_manager = CacheManager() 

File: backend/hono/src/core/cache.ts
/**
 * Redis缓存操作核心模块
 */

import { createClient, RedisClientType } from 'redis';
import { settings } from '../config';

class CacheManager {
  private client: RedisClientType | null = null;

  async init(): Promise<void> {
    try {
      this.client = createClient({
        url: settings.services.redisUrl,
      });

      this.client.on('error', (err) => {
        console.error('Redis连接错误:', err);
      });

      await this.client.connect();
      console.log('Redis连接成功');
    } catch (error) {
      console.error('Redis初始化失败:', error);
    }
  }

  async close(): Promise<void> {
    if (this.client) {
      await this.client.quit();
      this.client = null;
    }
  }

  async get<T = any>(key: string): Promise<T | null> {
    try {
      if (!this.client) return null;
      
      const data = await this.client.get(key);
      return data ? JSON.parse(data) : null;
    } catch (error) {
      console.error('Redis读取失败:', error);
      return null;
    }
  }

  async set<T = any>(key: string, value: T, ttl?: number): Promise<void> {
    try {
      if (!this.client) return;
      
      const serialized = JSON.stringify(value);
      const expiry = ttl || settings.cache.ttl;
      
      await this.client.setEx(key, expiry, serialized);
    } catch (error) {
      console.error('Redis写入失败:', error);
    }
  }

  async del(key: string): Promise<void> {
    try {
      if (!this.client) return;
      await this.client.del(key);
    } catch (error) {
      console.error('Redis删除失败:', error);
    }
  }

  async delPattern(pattern: string): Promise<void> {
    try {
      if (!this.client) return;
      
      const keys = await this.client.keys(pattern);
      if (keys.length > 0) {
        await this.client.del(keys);
      }
    } catch (error) {
      console.error('Redis批量删除失败:', error);
    }
  }

  async ping(): Promise<boolean> {
    try {
      if (!this.client) return false;
      const result = await this.client.ping();
      return result === 'PONG';
    } catch {
      return false;
    }
  }

  async getInfo(section: string = 'memory'): Promise<Record<string, any>> {
    try {
      if (!this.client) return {};
      
      const info = await this.client.info(section);
      const result: Record<string, any> = {};
      
      info.split('\r\n').forEach(line => {
        if (line.includes(':')) {
          const [key, value] = line.split(':');
          result[key] = value;
        }
      });
      
      return result;
    } catch (error) {
      console.error('获取Redis信息失败:', error);
      return {};
    }
  }
}

export const cacheManager = new CacheManager(); 